sync/src/block_sync.rs

```diff
@@ -0,0 +1,485 @@
+						}
+						None => {
+							debug!(target: "sync", "Could not revert to previous block, last: {} ({})", self.last_imported_block, self.last_imported_hash);
+						}
+					}
+				}
+			},
+			_ => (),
+		}
+		self.imported_this_round = None;
+	}
+
+	/// Find some headers or blocks to download for a peer.
+	pub fn request_blocks(&mut self, io: &mut SyncIo) -> Option<BlockRequest> {
+		match self.state {
+			State::Idle => {
+				self.start_sync_round(io);
+				return self.request_blocks(io);
+			},
+			State::ChainHead => {
+				// Request subchain headers
+				trace!(target: "sync", "Starting sync with better chain");
+				// Request MAX_HEADERS_TO_REQUEST - 2 headers apart so that
+				// MAX_HEADERS_TO_REQUEST would include headers for neighbouring subchains
+				return Some(BlockRequest::Headers {
+					start: self.last_imported_hash.clone(),
+					count: SUBCHAIN_SIZE,
+					skip: (MAX_HEADERS_TO_REQUEST - 2) as u64,
+				});
+			},
+			State::Blocks => {
+				// check to see if we need to download any block bodies first
+				let needed_bodies = self.blocks.needed_bodies(MAX_BODIES_TO_REQUEST, false);
+				if !needed_bodies.is_empty() {
+					return Some(BlockRequest::Bodies {
+						hashes: needed_bodies,
+					});
+				}
+
+				if self.download_receipts {
+					let needed_receipts = self.blocks.needed_receipts(MAX_RECEPITS_TO_REQUEST, false);
+					if !needed_receipts.is_empty() {
+						return Some(BlockRequest::Receipts {
+							hashes: needed_receipts,
+						});
+					}
+				}
+
+				// find subchain to download
+				if let Some((h, count)) = self.blocks.needed_headers(MAX_HEADERS_TO_REQUEST, false) {
+					return Some(BlockRequest::Headers {
+						start: h,
+						count: count as u64,
+						skip: 0,
+					});
+				}
+			},
+			State::Complete => (),
+		}
+		None
+	}
+
+	/// Checks if there are blocks fully downloaded that can be imported into the blockchain and does the import.
+	pub fn collect_blocks(&mut self, io: &mut SyncIo, allow_out_of_order: bool) -> Result<(), BlockDownloaderImportError> {
+		let mut bad = false;
+		let mut imported = HashSet::new();
+		let blocks = self.blocks.drain();
+		let count = blocks.len();
+		for block_and_receipts in blocks {
+			let block = block_and_receipts.block;
+			let receipts = block_and_receipts.receipts;
+			let (h, number, parent) = {
+				let header = BlockView::new(&block).header_view();
+				(header.sha3(), header.number(), header.parent_hash())
+			};
+
+			// Perform basic block verification
+			if !Block::is_good(&block) {
+				debug!(target: "sync", "Bad block rlp {:?} : {:?}", h, block);
+				bad = true;
+				break;
+			}
+
+			if self.target_hash.as_ref().map_or(false, |t| t == &h) {
+				self.state = State::Complete;
+				trace!(target: "sync", "Sync target reached");
+				return Ok(());
+			}
+
+			let result = if let Some(receipts) = receipts {
+				io.chain().import_block_with_receipts(block, receipts)
```

Comment:
while it can't affect the state, it's probably best not to ever propagate potentially invalid blocks. other peers might detect that and label the offending node as bad